gcd/src/main/java/org/spine3/server/storage/datastore/DsEventRecordStorage.java

```diff
@@ -63,31 +68,54 @@ protected DsEventRecordStorage(DsRecordStorage<EventId> storage) {
 
     @Override
     protected Map<EventId, EntityRecord> readRecords(EventStreamQuery query) {
-        final StructuredQuery<Entity> entityQuery = buildQuery(query);
+        final Iterable<StructuredQuery<Entity>> queries = buildQueries(query);
         final DatastoreWrapper datastore = getDelegateStorage().getDatastore();
-        final List<Entity> readEntities = datastore.read(entityQuery);
-        if (readEntities.isEmpty()) {
-            return Collections.emptyMap();
+        final Map<EventId, EntityRecord> result = new HashMap<>();
+        for (StructuredQuery<Entity> entityQuery : queries) {
+            final List<Entity> readEntities = datastore.read(entityQuery);
+            dumpResultsTo(result, readEntities);
         }
-        final Map<EventId, EntityRecord> result = processResults(readEntities);
         return result;
     }
 
-    private static StructuredQuery<Entity> buildQuery(EventStreamQuery sourceQuery) {
+    private static Iterable<StructuredQuery<Entity>> buildQueries(EventStreamQuery sourceQuery) {
         final Kind kind = Kind.of(TypeName.of(Event.class));
-        final Filter filter = buildFilter(sourceQuery);
+        final Filter timeFilter = buildTimeFilter(sourceQuery);
         final StructuredQuery.Builder<Entity> entityQuery = Query.newEntityQueryBuilder()
                                                                  .setKind(kind.getValue());
-        if (filter == null) {
-            return entityQuery.build();
+        final Collection<Filter> typeFilters = buildTypeFilters(sourceQuery);
+        if (typeFilters.isEmpty()) {
+            if (timeFilter != null) {
+                entityQuery.setFilter(timeFilter);
+            }
+            return singleton(entityQuery.build());
         } else {
-            return entityQuery.setFilter(filter)
-                              .build();
+            final Collection<StructuredQuery<Entity>> result = new LinkedList<>();
```

Comment:
Please use `newLinkedList()` for consistency with our code style.

Same applies to the similar statement below.